sql/api/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -66,37 +70,46 @@ object JavaTypeInference {
   private def encoderFor(
       t: Type,
       seenTypeSet: Set[Class[_]],
-      typeVariables: Map[TypeVariable[_], Type] = Map.empty): AgnosticEncoder[_] = t match {
-
-    case c: Class[_] if c == java.lang.Boolean.TYPE => PrimitiveBooleanEncoder
-    case c: Class[_] if c == java.lang.Byte.TYPE => PrimitiveByteEncoder
-    case c: Class[_] if c == java.lang.Short.TYPE => PrimitiveShortEncoder
-    case c: Class[_] if c == java.lang.Integer.TYPE => PrimitiveIntEncoder
-    case c: Class[_] if c == java.lang.Long.TYPE => PrimitiveLongEncoder
-    case c: Class[_] if c == java.lang.Float.TYPE => PrimitiveFloatEncoder
-    case c: Class[_] if c == java.lang.Double.TYPE => PrimitiveDoubleEncoder
-
-    case c: Class[_] if c == classOf[java.lang.Boolean] => BoxedBooleanEncoder
-    case c: Class[_] if c == classOf[java.lang.Byte] => BoxedByteEncoder
-    case c: Class[_] if c == classOf[java.lang.Short] => BoxedShortEncoder
-    case c: Class[_] if c == classOf[java.lang.Integer] => BoxedIntEncoder
-    case c: Class[_] if c == classOf[java.lang.Long] => BoxedLongEncoder
-    case c: Class[_] if c == classOf[java.lang.Float] => BoxedFloatEncoder
-    case c: Class[_] if c == classOf[java.lang.Double] => BoxedDoubleEncoder
-
-    case c: Class[_] if c == classOf[java.lang.String] => StringEncoder
-    case c: Class[_] if c == classOf[Array[Byte]] => BinaryEncoder
-    case c: Class[_] if c == classOf[java.math.BigDecimal] => DEFAULT_JAVA_DECIMAL_ENCODER
-    case c: Class[_] if c == classOf[java.math.BigInteger] => JavaBigIntEncoder
-    case c: Class[_] if c == classOf[java.time.LocalDate] => STRICT_LOCAL_DATE_ENCODER
-    case c: Class[_] if c == classOf[java.sql.Date] => STRICT_DATE_ENCODER
-    case c: Class[_] if c == classOf[java.time.Instant] => STRICT_INSTANT_ENCODER
-    case c: Class[_] if c == classOf[java.sql.Timestamp] => STRICT_TIMESTAMP_ENCODER
-    case c: Class[_] if c == classOf[java.time.LocalDateTime] => LocalDateTimeEncoder
-    case c: Class[_] if c == classOf[java.time.Duration] => DayTimeIntervalEncoder
-    case c: Class[_] if c == classOf[java.time.Period] => YearMonthIntervalEncoder
-
-    case c: Class[_] if c.isEnum => JavaEnumEncoder(ClassTag(c))
+      typeVariables: Map[TypeVariable[_], Type] = Map.empty,
+      forGenericBound: Boolean = false): AgnosticEncoder[_] = t match {
+
+    case c: Class[_] if !forGenericBound && c == java.lang.Boolean.TYPE => PrimitiveBooleanEncoder
```

Comment:
For readability purposes I would create a branch at the beginning where you handle `case tv: TypeVariable[_] if forGenericBound =>` This way the rest of the code is less impacted.

Reply:
I have sort of done it now..